# Task
Read a supervised dataset and use bagging and boosting techniques to classify the dataset. Indicate the performance of the models.

## Load data

### Subtask:
Load the supervised dataset into a pandas DataFrame.


**Reasoning**:
Import pandas and load the dataset into a DataFrame, then display the first few rows.



In [13]:
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load the breast cancer dataset
breast_cancer = load_breast_cancer()

# Convert to a pandas DataFrame
df = pd.DataFrame(data=breast_cancer.data, columns=breast_cancer.feature_names)

# Add the target variable to the DataFrame
df['target'] = breast_cancer.target

# Display the first few rows and info
display(df.head())
df.info()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

## Preprocess data

### Subtask:
Handle missing values, encode categorical features, and split the data into training and testing sets.


**Reasoning**:
Check for missing values, separate features and target, and split the data into training and testing sets as per the instructions.



In [14]:
# 1. Check for missing values
print("Missing values per column:")
print(df.isnull().sum())

# 2. Identify categorical columns - Not necessary as per instructions, all features are numerical and target is numerical.

# 3. Separate features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# 4. Split data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("\nData splitting complete.")
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

Missing values per column:
mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
target                     0
dtype: int64

Data splitting complete.
X_train shape: (455, 30)
X_test sha

## Implement bagging

### Subtask:
Train a bagging classifier on the training data.


**Reasoning**:
Import the BaggingClassifier and train it on the training data.



In [15]:
from sklearn.ensemble import BaggingClassifier

# Instantiate a BaggingClassifier object with default parameters
bagging_clf = BaggingClassifier(random_state=42)

# Train the bagging classifier on the training data
bagging_clf.fit(X_train, y_train)

print("BaggingClassifier trained successfully.")

BaggingClassifier trained successfully.


## Implement boosting

### Subtask:
Train a boosting classifier (e.g., AdaBoost or Gradient Boosting) on the training data.


**Reasoning**:
Train a boosting classifier on the training data using Gradient Boosting.



In [16]:
from sklearn.ensemble import GradientBoostingClassifier

# Instantiate a GradientBoostingClassifier object with default parameters
# and set random_state for reproducibility
boosting_clf = GradientBoostingClassifier(random_state=42)

# Train the boosting classifier on the training data
boosting_clf.fit(X_train, y_train)

print("Boosting classifier trained successfully.")

Boosting classifier trained successfully.


## Evaluate models

### Subtask:
Evaluate the performance of both the bagging and boosting models on the testing data using appropriate metrics (e.g., accuracy, precision, recall, F1-score).


**Reasoning**:
Evaluate the performance of both the bagging and boosting models on the testing data using accuracy, precision, recall, and F1-score.



In [17]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Evaluate Bagging Classifier
bagging_predictions = bagging_clf.predict(X_test)
bagging_accuracy = accuracy_score(y_test, bagging_predictions)
bagging_precision = precision_score(y_test, bagging_predictions)
bagging_recall = recall_score(y_test, bagging_predictions)
bagging_f1 = f1_score(y_test, bagging_predictions)

print("Bagging Classifier Performance:")
print(f"Accuracy: {bagging_accuracy:.4f}")
print(f"Precision: {bagging_precision:.4f}")
print(f"Recall: {bagging_recall:.4f}")
print(f"F1-score: {bagging_f1:.4f}")

print("-" * 30)

# Evaluate Boosting Classifier
boosting_predictions = boosting_clf.predict(X_test)
boosting_accuracy = accuracy_score(y_test, boosting_predictions)
boosting_precision = precision_score(y_test, boosting_predictions)
boosting_recall = recall_score(y_test, boosting_predictions)
boosting_f1 = f1_score(y_test, boosting_predictions)

print("Boosting Classifier Performance:")
print(f"Accuracy: {boosting_accuracy:.4f}")
print(f"Precision: {boosting_precision:.4f}")
print(f"Recall: {boosting_recall:.4f}")
print(f"F1-score: {boosting_f1:.4f}")

Bagging Classifier Performance:
Accuracy: 0.9561
Precision: 0.9583
Recall: 0.9718
F1-score: 0.9650
------------------------------
Boosting Classifier Performance:
Accuracy: 0.9561
Precision: 0.9583
Recall: 0.9718
F1-score: 0.9650


## Compare performance

### Subtask:
Compare the performance of the bagging and boosting models and summarize the results.


**Reasoning**:
Summarize the performance comparison based on the calculated metrics.



In [18]:
print("Performance Comparison:")
print("Bagging Classifier:")
print(f"  Accuracy: {bagging_accuracy:.4f}")
print(f"  Precision: {bagging_precision:.4f}")
print(f"  Recall: {bagging_recall:.4f}")
print(f"  F1-score: {bagging_f1:.4f}")

print("\nBoosting Classifier:")
print(f"  Accuracy: {boosting_accuracy:.4f}")
print(f"  Precision: {boosting_precision:.4f}")
print(f"  Recall: {boosting_recall:.4f}")
print(f"  F1-score: {boosting_f1:.4f}")

print("\nSummary:")
if bagging_accuracy == boosting_accuracy and bagging_precision == boosting_precision and bagging_recall == boosting_recall and bagging_f1 == boosting_f1:
    print("Both Bagging and Boosting classifiers achieved identical performance across all evaluated metrics (Accuracy, Precision, Recall, and F1-score) on the test set.")
else:
    print("Comparing the performance metrics:")
    print(f"- Accuracy: Bagging ({bagging_accuracy:.4f}) vs Boosting ({boosting_accuracy:.4f})")
    print(f"- Precision: Bagging ({bagging_precision:.4f}) vs Boosting ({boosting_precision:.4f})")
    print(f"- Recall: Bagging ({bagging_recall:.4f}) vs Boosting ({boosting_recall:.4f})")
    print(f"- F1-score: Bagging ({bagging_f1:.4f}) vs Boosting ({boosting_f1:.4f})")
    print("Based on these metrics, you can compare which model performed better.")

Performance Comparison:
Bagging Classifier:
  Accuracy: 0.9561
  Precision: 0.9583
  Recall: 0.9718
  F1-score: 0.9650

Boosting Classifier:
  Accuracy: 0.9561
  Precision: 0.9583
  Recall: 0.9718
  F1-score: 0.9650

Summary:
Both Bagging and Boosting classifiers achieved identical performance across all evaluated metrics (Accuracy, Precision, Recall, and F1-score) on the test set.


## Summary:

### Data Analysis Key Findings

*   The initial attempt to load a custom supervised dataset failed because the file was not found.
*   The Breast Cancer dataset was successfully loaded from `sklearn.datasets` and used for the task.
*   The dataset contained no missing values, and all features were numerical, eliminating the need for imputation or categorical encoding.
*   The dataset was split into training (80%) and testing (20%) sets, resulting in 455 training samples and 114 testing samples.
*   Both a Bagging Classifier and a Gradient Boosting Classifier were successfully trained on the training data.
*   Both the Bagging and Boosting classifiers achieved identical performance metrics on the test set:
    *   Accuracy: 0.9561
    *   Precision: 0.9583
    *   Recall: 0.9718
    *   F1-score: 0.9650

### Insights or Next Steps

*   Given the identical performance, further investigation could involve hyperparameter tuning for both models to see if either technique can achieve superior results.
*   Exploring other ensemble techniques or base estimators might be beneficial to potentially improve performance or understand the dataset characteristics better.
